### PR TITLE
34 complete db functions

### LINKED ISSUE #34: Complete_DB_Functions
Need to create the following functions and post/put functions:
1. Clean_Preq
2. Marked_Preq
3. Dependencies

It will likely be helpful (and necessary) to create helper get function from DB.

### PR DESCRIPTION
**Pull request by fraser-mcleod**
Completed all course database post and put functions including clean_preq, marked_preq, and dependent_courses. Some things to consider before moving forward:
1. Might be better to store dependencies differently since they are simply just a list
2. There are many courses mentioned in preqs that are no longer offered and therefore do not have an entry in the course table. We should think about what we want to do with these.

### CODE REVIEW

File: server/webscraping/postCleanPreq.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    return courseList
+
+def createCleanPreq(raw_preq: str):
+    """
+    Given a a courses raw_preq, return its clean_preq
+    :return clean_preq - cleaned string that makes all the course_codes explicit
+    """
+    form = Form(raw_preq)
+    clean_preq = form.formalizeCourseNames()
```
Comment (marmikpatel19) [inline server/webscraping/postCleanPreq.py:40]: Why are we only using this function and not the others ones we previously had. Are those unnecessary?

File: server/webscraping/postCleanPreq.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    fullCourseRegEx = '([A-Z]{2,4}\s[0-9]{3})'
+    before_marker = '$['
+    after_marker = ']$'
+    replacement = before_marker + '\g<1>' + after_marker
+
+    marked_preq = re.sub(fullCourseRegEx, replacement, clean_preq)
+
+    return marked_preq
+
+
```
Comment (marmikpatel19) [inline server/webscraping/postCleanPreq.py:57]: Why the spaces between all these functions?

File: server/webscraping/postDependentCourses.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import requests
+import environ
+
+env = environ.Env()
+environ.Env.read_env()
+# Steps:
+# 1. Need to get all courses from DB ans store in list
```
Comment (marmikpatel19) [inline server/webscraping/postDependentCourses.py:7]: typo at "ans"

File: server/webscraping/postDependentCourses.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+# 2. We will need a dictionary with course codes as keys and dependent course list as values, initially all empty
+# 3. For each course c:
+#       - Use clean/marked preq to find all of its preq courses p. 
+#       - For each of those preq courses p, append c to its dependancy list using dictionary from step 2
```
Comment (marmikpatel19) [inline server/webscraping/postDependentCourses.py:11]: Great algorithm!

**Review by marmikpatel19 — COMMENTED** [review]
Looks good--great work! Left some comments.

If you want to create an issue for it, I can take care of the handling for the non-existent courses.

File: server/webscraping/postCleanPreq.py (hunk 1 above)
Comment (fraser-mcleod) [reply server/webscraping/postCleanPreq.py:40]: The other ones were to convert forms all into exact conjunctive normal form - which was quite difficult and complicated. Now, we just want to make sure all course codes are explicit in the preq string. Thus, unnecessary for current application, but, they may be helpful in the future.

File: server/webscraping/postCleanPreq.py (hunk 2 above)
Comment (fraser-mcleod) [reply server/webscraping/postCleanPreq.py:57]: No reason for the spaces

File: server/webscraping/postDependentCourses.py (hunk 4 above)
Comment (fraser-mcleod) [reply server/webscraping/postDependentCourses.py:11]: :)

**Pull request merged**